thegamecracks/thegamebot
# Feature Update - Help command, Reminders

* Create dbsetup.py for managing database setups

- Removes setup function main() in database.py
- Moves Users and Notes table SQL from database.py into their corresponding scripts
- Removes database path from database.py
- Creates dbsetup.py which contains database connection instances from database.py and setup functions
- Updates main to use `database.dbsetup.setup()`
- `dbconnection_users` renamed to `dbconn_users`

* Minor changes and refactoring

- Returns dmtest command to original functionality
- Changes connect event handlers and `background.print_timestamp` to display local time
- Refactors error handlers to extract original error in one line, reducing indentation

* Transition database to using aiosqlite

- Switches database library from sqlite3 to aiosqlite to allow concurrent database usage
- Renames `UserDatabase.__contains__` to async `UserDatabase.has_user`

* Update error handlers + tweaks

- Adds `on_command_error` event handler which handles almost every subclass of CommandError
- Retires `utils.print_error` decorator as it intercepts `MissingRequiredArgument` and `CommandInvokeError` from reaching actual error handlers
- Several changes to command error handlers in preparation for fully handling command errors
- Changes default prefix to "."

* Add custom help command + update commands + update event handlers

- Adds help command with full support for individual commands, group commands, non-interactive paginated cog commands, and non-interactive paginated category lists
- All commands now have cooldowns
- Moves event handlers (on_connect, on_command_error, etc) into bot/eventhandlers.py to be registered in main.py using `eventhandlers.setup(bot)`
- The bot color has been moved into settings.json instead of being a constant
- Adds number limits onto some mathematic commands
- Fixes getlastmessage command not escaping mentions
- Increases coinflip and dice commands' response delay from 1 to 1.5 seconds

* New command + Fixes + Refactoring

- Default prefix changed to semi-colon;
- New "hyperlink" command group allows server members to create hyperlinks with custom display text
- "execute" command now supports code blocks
- "userinfo" no longer reraises unspecified errors
- EmbedTools cog (and file) renamed to Embedding as it prevented the help command from recognizing it
- Fixes "getlastmessage" not returning early if the given channel isn't a TextChannel
- Removes unnecessary returns from error handlers (remnants of old code)
- All cases of `await ... \n return` in commands have been replaced with `return await ...` to reduce line count

* Add custom prefixes + two admin commands

- Database now stores custom prefixes for guilds and can be modified with the "changeprefix" command for anyone with manage guilds permissions
  - Prefixes are limited to 20 chars
  - Tested against SQL injection (although I'm not a pen tester)
  - get_prefix() function located in bot/database/dbsetup.py
- "prefix" setting in settings.json changed to "default_prefix"
- Adds support for calling commands by mentioning it
- bot.settings now caches its reads and is cleared when `overwrite_setting`, `update_settings`, or `write_setting` is called or by using the new `clear_cache()` function
- Adds three admin commands
  - "block" command hangs the bot for a specified amount of time by using `time.sleep()`
  - "cooldown" command resets the cooldown for a given command (only for the caller)
  - "clearsettingscache" command clears the cache for settings
- Fixes help command failing when a command has no description
- Fixes not being able to get help for Undefined category by renaming NoCategory cog (and the file) to Undefined
- Moves "dmtest" and "test" commands into Undefined category
- `Database` class now uses slots

* Update several commands and on_command_error handler

- "execute": Adds discord library to environment, id is now printed in terminal instead of mention
- "presence": Refactors displaying unknown subcommand, fixes having two error messages sent for `CommandNotFound` and `MissingRequiredArgument`
- Random Presence loop: Fixes bug where it mutates the settings cache and causes an AttributeError when using a presence a second time
- bot.utils.datetime_difference_string now uses `inflect` library to join the time (adding "and" at the end)
- "changeprefix": Allows the bot owner to also change the prefix regardless of permissions
- on_command_error:
  - Cleans up command signature to no longer show type annotations
  - Missing permissions are displayed with proper words instead of underscore-separated text

* Bugfixes related to guild-only commands

- Adds a response when a user tries to use a guild only command (raising NoPrivateMessage)
- Refactors "serverinfo" command to use `commands.guild_only()` check instead of an in-code handler

* Bugfixes + update discord.py to 1.5.1

- "userinfo" command:
  - Adds hardcoded option for allowing server members to be queried in DMs
  - Fixes displaying two "cannot find user" error messages when the member can't be found
  - Handles specific cases related to members intent and DM/server invokation
  - Fixes displaying member status as Offline when either presences or members intent are disabled
- on_command_error now prints which command the user attempted using
- Removes a testing attribute from the Bot instance
- Updates discord.py to 1.5.1 in requirements.txt

* Add presence subcommand + update background cog

- Adds "presence competing" subcommand to allow setting the bot's presence with the "competing" activity type (new in d.py v1.5)
- BackgroundTasks cog (now Tasks cog)
  - Adds random presence support for competing activity type
  - Refactors extracting status to remove the need to copy presence dict
- Adds 5 competing random presences to settings.json and compacts every presence
- Updates settings.py to include new presences

* Add reminder system

- New reminder system that allows users to add their own reminders, which the bot will then send messages on the due date
  - Uses database caching of users and their reminders
  - Tasks are created every 10 minutes for any reminder that is about to be sent within 11 minutes or is overdue
  - Dependent on bot/classes/timeobj.py
  - This cog requires members intent to send DMs, therefore tasks will not be created and messages sent from commands include a disclaimer (although adding/removing/showing reminders will work just fine)
- Adds database caching for Notes
- `Database.add_row` now returns the last row ID and is propagated through the `add_*` methods of subclasses
- bot.utils.datetime_difference_string can now display hours and seconds, and allows specifying which units of time to display
  - Informative commands tweaked to not display seconds
- Some bugfixes for random presences
